packages/embark/src/cmd/cmd_controller.js

```diff
@@ -210,25 +213,31 @@ class EmbarkController {
           engine.events.emit("status", __("Ready").green);
         });
 
-        engine.events.on('file-event', async ({ fileType, path }) => {
-          // TODO: re-add async.cargo / or use rxjs to use latest request in the queue
-          console.dir("-- before timeout - file changed")
-
-          if (fileType === 'contract' || fileType === 'config') {
-            let contractsFiles = await engine.events.request2("config:contractsFiles");
-            let compiledContracts = await engine.events.request2("compiler:contracts:compile", contractsFiles);
-            let _contractsConfig = await engine.events.request2("config:contractsConfig");
-            let contractsConfig = cloneDeep(_contractsConfig);
-            let [contractsList, contractDependencies] = await engine.events.request2("contracts:build", contractsConfig, compiledContracts);
-            await engine.events.request2("deployment:contracts:deploy", contractsList, contractDependencies);
-          } else if (fileType === 'asset') {
-            engine.events.request('pipeline:generateAll', () => {
-              console.dir("outputDone")
-              engine.events.emit('outputDone');
-            });
-          }
+        const fileChanges$ = Observable.create(obs => {
+          engine.events.on('file-event', (data) => {
+            obs.next(data);
+          });
         });
 
+        fileChanges$.pipe(
+          exhaustMapWithLastIgnored(({ fileType, path }) => {
+            if (fileType === 'contract' || fileType === 'config') {
+              const contractFiles$ = from(engine.events.request2("config:contractsFiles"));
+              const compiledContracts$ = contractFiles$.pipe(
+                mergeMap(contractsFiles => from(engine.events.request2("compiler:contracts:compile", contractsFiles)))
+              );
+              const contractsConfig$ = from(engine.events.request2("config:contractsConfig"));
+
+              return forkJoin(contractsConfig$, compiledContracts$).pipe(
+                mergeMap(([contractsConfig, compiledContracts]) => from(engine.events.request2("contracts:build", contractsConfig, compiledContracts))),
+                mergeMap(([contractsList, contractDependencies]) => from(engine.events.request2("deployment:contracts:deploy", contractsList, contractDependencies)))
+              );
+            } else if (fileType === 'asset') {
+              return from(engine.events.request2('pipeline:generateAll')).pipe(tap(_ => engine.events.emit('outputDone')));
+            }
+          })
+        ).subscribe();
```

Comment:
Notice that the actual work happens inside the pipe instead of the subscribe block. The reason for that is that we first have to modify the stream in a certain way (ignore `fileChanges$` while a compilation/deployment is running, and only process the last "queued" emission). It just so happens that the stream we're looking for requires us to do the work that we want to do anyways.

Also notice that this is a long-living observable (it never completes), however we don't have to unsubscribe here because there won't be multiple `run()` calls at a time.